core/src/socket.rs

```diff
@@ -111,100 +160,238 @@ impl<'gc> Sockets<'gc> {
+                                .classes()
+                                .progressevent
+                                .construct(
+                                    &mut activation,
+                                    &[
+                                        "socketData".into(),
+                                        false.into(),
+                                        false.into(),
+                                        bytes_loaded.into(),
+                                        //NOTE: bytesTotal is not used by socketData event.
+                                        0.into(),
+                                    ],
+                                )
+                                .expect("ProgressEvent should be constructed");
+
+                            Avm2::dispatch_event(
+                                &mut activation.context,
+                                progress_evt,
+                                target.into(),
+                            );
+                        }
+                        SocketKind::Avm1(target) => {
+                            let mut activation = Avm1Activation::from_stub(
+                                context.reborrow(),
+                                ActivationIdentifier::root("[XMLSocket]"),
+                            );
+
+                            // NOTE: This is enforced in connect_avm1() function.
+                            let xml_socket =
+                                XmlSocket::cast(target.into()).expect("target should be XmlSocket");
+
+                            // Get XmlSocket read buffer.
+                            let mut buffer = xml_socket.read_buffer();
+                            buffer.extend(data);
+
+                            let mut messages = vec![];
+
+                            // Check for messages.
+                            while let Some((index, _)) =
+                                buffer.iter().enumerate().find(|(_, &b)| b == 0)
+                            {
+                                let message = buffer.drain(..index).collect::<Vec<_>>();
+                                // Remove null byte.
+                                let _ = buffer.drain(..1);
+
+                                // Store the message in cache.
+                                messages.push(AvmString::new_utf8_bytes(activation.gc(), &message));
+                            }
+
+                            // Drop the reference to the buffer, to make sure the SWF can call with close.
+                            drop(buffer);
+
+                            for message in messages {
+                                let _ = target.call_method(
+                                    "onData".into(),
+                                    &[message.into()],
+                                    &mut activation,
+                                    ExecutionReason::Special,
+                                );
+                            }
```

Comment:
If the socket gets closed when multiple messages are still pending in `messages`, we'll still call `onData` even if user's code asked to close the socket.

```suggestion
                            for message in messages {
                                let _ = target.call_method(
                                    "onData".into(),
                                    &[message.into()],
                                    &mut activation,
                                    ExecutionReason::Special,
                                );
                                
                                // Drop remaining messages if the socket was closed.
                                if !context.sockets.sockets.contains(handle) {
                                    break;
                                }
                            }
```

Reply:
Not sure what Flash's behavior here is at this point, but I think we can leave it like this for now and change it in the future if it becomes a problem.